### PR TITLE
Add cipher suite tests to ssl_test_old for DTLS

### PR DESCRIPTION
**Pull request by dwmw2**
As [discussed](https://github.com/openssl/openssl/pull/1666#issuecomment-252048193), `ssl_test_old` iterates over all ciphersuites for TLS protocols but doesn't do DTLS.

I am no longer too bothered about that addition, since in #1666 I've since added a separate MTU test which will iterate over all DTLS ciphersuites anyway. But in order to get `ssl_test_old` doing it, I also had to fix `SSL_CTX_set_cipher_list()` to accept "`DTLSv1`" and "`DTLSv1.2`" as arguments. That might be worth having, so I'm submitting it separately for posterity.

### CODE REVIEW

File: test/recipes/80-test_ssl_old.t [hunk 1]
```diff
@@ -466,10 +468,15 @@ sub testssl {
 	plan tests => $protocolciphersuitcount + scalar(@protocols);
 
 	foreach my $protocol (@protocols) {
+	    my $protoarg = "";
+	    $protoarg = "-ssl3" if $protocol eq "SSLv3";
+	    $protoarg = "-dtls1" if $protocol eq "DTLSv1";
+	    $protoarg = "-dtls12" if $protocol eq "DTLSv1.2";
```
Comment (levitte) [inline test/recipes/80-test_ssl_old.t]: I would very much prefer if you made `protoarg` a list.  The reason is that the default value will give a 0 char argument string, and there's a risk that this becomes an actual empty argument in the call below, if it gets quoted.  It's better if the default is like this, as that will just disappear:

``` perl
	    my @protoarg = ();
```

File: test/recipes/80-test_ssl_old.t [hunk 2]
```diff
@@ -466,10 +468,15 @@ sub testssl {
...
+
 	    note "Testing ciphersuites for $protocol";
 	    foreach my $cipher (@{$ciphersuites{$protocol}}) {
 		ok(run(test([@ssltest, @exkeys, "-cipher", $cipher,
-			     $protocol eq "SSLv3" ? ("-ssl3") : ()])),
+			     ($protoarg)])),
```
Comment (levitte) [inline test/recipes/80-test_ssl_old.t]: So this should be changed to:

``` perl
			     @protoarg])),
```

Comment (dwmw2) [inline test/recipes/80-test_ssl_old.t]: I have no idea what you just said.  Are you assuming I know perl? :)

Will attempt this.

**Comment by richsalz** [comment]
I think with all those parens it’s more like lisp :)

**Comment by dwmw2** [comment]
> As for testing in general, we would prefer if new tests are added to 80-test_ssl_new.t rather than the old.

This isn't exactly a *new* test... it is existing code in the old test, which *already* supported DTLS. It was just never *invoked* before... just because we had no way to list the ciphersuites with `openssl ciphers DTLSv1` ... which I now added. 

I *have* also added a new test which does this, in #1666.

File: test/recipes/80-test_ssl_old.t (hunk 2 above)
Comment (levitte) [inline test/recipes/80-test_ssl_old.t]: So this specific line that I'm commenting here, just change exactly like I showed you.

File: test/recipes/80-test_ssl_old.t (hunk 1 above)
Comment (levitte) [inline test/recipes/80-test_ssl_old.t]: Replace those four lines with this:

``` perl
	    my @protoarg = ();
 	    @protoarg = ("-ssl3") if $protocol eq "SSLv3";
 	    @protoarg = ("-dtls1") if $protocol eq "DTLSv1";
 	    @protoarg = ("-dtls12") if $protocol eq "DTLSv1.2";
```

File: test/recipes/80-test_ssl_old.t [hunk 3]
```diff
@@ -466,10 +468,15 @@ sub testssl {
 	plan tests => $protocolciphersuitcount + scalar(@protocols);
 
 	foreach my $protocol (@protocols) {
+	    my @protoarg = ();
+	    @protoarg = ("-ssl3") if $protocol eq "SSLv3";
+	    @protoarg = ("-dtls1") if $protocol eq "DTLSv1";
+	    @protoarg = ("-dtls12") if $protocol eq "DTLSv1.2";
+
```
Comment (mattcaswell) [inline test/recipes/80-test_ssl_old.t]: There is a non-trivial conflict in this section with current master

File: apps/ciphers.c [hunk 4]
```diff
@@ -135,6 +143,20 @@ int ciphers_main(int argc, char **argv)
             min_version = TLS1_2_VERSION;
             max_version = TLS1_2_VERSION;
             break;
+#ifndef OPENSSL_NO_DTLS1
```
Comment (mattcaswell) [inline apps/ciphers.c]: I don't think these guards are actually necessary here. We don't have them for the TLS options.

**Comment by mattcaswell** [comment]
Just taking away the "reviewed" label for the moment...due to the above issues.

**Comment by richsalz** [comment]
okay from me with or without the guards.  @mattcaswell still want them removed?
+1

**Comment by richsalz** [comment]
and since it's a test, +1 for 1.1.0 too.

**Comment by mattcaswell** [comment]
> okay from me with or without the guards. @mattcaswell still want them removed?

I'd rather avoid cluttering the code up with guards that we don't need. In any case this branch needs to have the merge conflicts resolved before we can progress.

**Comment by mattcaswell** [comment]
Ping @dwmw2. This seems to have stalled. Are you planning on progressing this?

**Comment by richsalz** [comment]
we don't really want things added to the old SSL test.

**Comment by mattcaswell** [comment]
This seems to have stalled beyond the point of resurrection. Closing.

**Pull request closed without merging**